spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -585,6 +625,54 @@ private void ackImmediate(ConsumerRecord<K, V> record) {
 		}
 
 		private void invokeListener(final ConsumerRecords<K, V> records) {
+			if (this.isBatchListener) {
+				invokeBatchListener(records);
+			}
+			else {
+				invokeRecordListener(records);
+			}
+		}
+
+		private void invokeBatchListener(final ConsumerRecords<K, V> records) {
+			List<ConsumerRecord<K, V>> recordList = new LinkedList<ConsumerRecord<K, V>>();
+			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
+			while (iterator.hasNext()) {
+				recordList.add(iterator.next());
+			}
+			if (recordList.size() > 0) {
+				try {
+					if (this.batchAcknowledgingMessageListener != null) {
+						this.batchAcknowledgingMessageListener.onMessage(recordList,
+								new ConsumerBatchAcknowledgment(recordList, this.isManualImmediateAck));
+					}
+					else {
+						this.batchListener.onMessage(recordList);
+						if (!this.isAnyManualAck && !this.autoCommit) {
+							for (ConsumerRecord<K, V> record : recordList) {
+								this.acks.put(record);
+							}
+						}
+					}
+				}
+				catch (Exception e) {
+					if (this.containerProperties.isAckOnError() && !this.autoCommit) {
+						for (ConsumerRecord<K, V> record : recordList) {
+							this.acks.add(record);
```

Comment:
My assumption has been done on this JavaDoc for `add()`:
```
 * Inserts the specified element into this queue if it is possible to do
     * so immediately without violating capacity restrictions, returning
     * {@code true} upon success and throwing an
     * {@code IllegalStateException} if no space is currently available.
     * When using a capacity-restricted queue, it is generally preferable to
     * use {@link #offer(Object) offer}.
```
That's why I say we don't need to worry about capacity for the `LinkedBlockingQueue`. I just wondered why a different method has been chosen...